internal/errors/is.go

```diff
@@ -89,3 +89,20 @@ func IsMissingTableError(err error) bool {
 	}
 	return false
 }
+
+// IsNotFoundError returns a boolean indicating whether the error is known to
+// report a not found violation.
+func IsNotFoundError(err error) bool {
```

Comment:
This function illustrates one of my concerns with the multi-Code model. Switching logic that switches on Code (via using functions like this) can have different behavior depending on the ordering in which the codes are tested in the switch. At the same time, this will return if any error in the chain is RecordNotFound, but not necessarily if the originating error is, or even the most recent error. That means behavior of a switch on Code is also dependent on changes to logic that might seem innocuous -- e.g. if one of the wrapped errors decides to add or modify the code it's wrapping, that can also affect the behavior of this function.

Reply:
The current implementation of this `IsX` functions only check if the most recent domain error's code is of a certain type not any in the chain, @mgaffney mentioned in the call yesterday that perhaps we should have the ability to check if any in the chain match which I did give some thought to but have not implemented yet

Reply:
Or we could just not support multiple Codes at this time and only add it in if/when needed. There was a question a few days ago about whether or not we can think of a time when we'd want to supply more than one Code. Did we ever think of any?

Reply:
So the only time we currently do this is when updating/creating a Host and we get a db `NotNull` or `ConstraintViolation` for both these cases the error returned becomes one with a code of `InvalidAddress`
https://github.com/hashicorp/boundary/blob/e8cf4885e5450b74bbcfd13b4a191dd1d40953e5/internal/host/static/repository_host.go#L178

Reply:
We might also do something similar when authenticating (not found error for auth method or login name could be translated into an unauthenticated error potentially)

Reply:
In that case don't we have a single Code anyways? This is the point where we're translating the DB errors to API errors. If two DB errors map to the same API Code, it's still a single Code right?

Reply:
Well the top level would have one code, and therefore the `IsX` function, as it is currently written, will only be satisfied if the top level is `X`.  But the wrapped errors could have a different code:
`{"Code": InvalidAddress, "Msg": "Top level message", "Wrapped": {"Code": "NotNull", "Msg": "Db level message"}}`

Reply:
Currently, the only time the wrapped errors are used are in the stringer